tool/templates/cpp/runtime.hpp.jinja

```diff
@@ -236,6 +249,54 @@ template <typename Ret, typename... Args> struct fn_traits<std::function<Ret(Arg
 template<class T>
 fn_traits(T) -> fn_traits<T>;
 
+
+// trait to detect if something is std::optional
+template<class T>
+inline constexpr bool is_optional_v = std::false_type{};
+
+template<class T>
+inline constexpr bool is_optional_v<std::optional<T>> = std::true_type{};
+
+// Trait for extracting inner types from either std::optional or std::unique_ptr.
+// These are the two potential types returned by next() functions
+template<typename T> struct inner { using type = void; };
+template<typename T> struct inner<std::unique_ptr<T>> { using type = T; };
+template<typename T> struct inner<std::optional<T>>{ using type = T; };
+
+// Adapter for iterator types
+template<typename T, typename U = void> struct has_next : std::false_type {};
+template<typename T> struct has_next < T, std::void_t<decltype(std::declval<T>().next())>> : std::true_type {};
+template<typename T> constexpr bool has_next_v = has_next<T>::value;
+
+/// Helper template enabling native iteration over unique ptrs to objects which implement next()
+template<typename T>
+struct next_to_iter_helper {
+  static_assert(has_next_v<T>, "next_to_iter_helper may only be used with types implementing next()");
+  using next_type = decltype(std::declval<T>().next());
+
+  // STL Iterator trait definitions
+  using value_type = typename inner<next_type>::type;
+  using difference_type = void;
+  using reference = std::add_lvalue_reference_t<value_type>;
+  using iterator_category = std::input_iterator_tag;
+
+  next_to_iter_helper(std::unique_ptr<T>&& ptr) : _ptr(std::move(ptr)), _curr(_ptr->next()) {}
+  
+  // https://en.cppreference.com/w/cpp/named_req/InputIterator requires that the type be copyable
+  next_to_iter_helper(const next_to_iter_helper& o) : _ptr(o._ptr), _curr(o._curr) {}
+
+  void operator++() { _curr = _ptr->next(); }
```

Comment:
note that structs and enums can also have iterators. that should still work since here they should be constructed as references

Reply:
It's going to be super rare to do this, so it's unimportant for now.